File: internal/buildinfo/buildinfo.go
// Package buildinfo holds build-time identity values: version, commit,
// build date, and the Application Insights connection string used for
// opt-out telemetry. Version, Commit and Date are populated at link
// time by goreleaser; AppInsightsConnString is a committed constant so
// every build — release, source, or fork — reports identically.
package buildinfo

// Version is the CalVer release string (e.g. "2026.05.1"). Set via
// -ldflags "-X github.com/sdebruyn/onelake-explorer-macos/internal/buildinfo.Version=...".
// In a plain `go build` it stays "dev".
var Version = "dev"

// Commit is the short git commit SHA. Optional; set via ldflags.
var Commit = ""

// Date is the ISO-8601 build timestamp. Optional; set via ldflags.
var Date = ""

// AppInsightsConnString is the Application Insights connection string
// every OFEM build reports to.
//
// It is deliberately a committed constant rather than a build-time
// secret. Per Microsoft's design and documentation, an Application
// Insights connection string is write-only and is meant to be public —
// the same string ships in every browser-side JS app, mobile binary,
// or desktop app that uses Application Insights. Committing it here
// does not change the security posture (any user can already extract
// it from the official cask release with `strings`) and it has one
// concrete upside: source-built binaries and forks participate in the
// shared opt-out telemetry stream too, so the maintainer gets a
// representative signal instead of only the Homebrew users.
//
// Users disable telemetry per install with `ofem config set telemetry
// off` or `OFEM_TELEMETRY=0`. There is no opt-in path that requires
// editing this file.
const AppInsightsConnString = "InstrumentationKey=bb7c05e2-4616-4b8d-a18a-e32128034eb4;IngestionEndpoint=https://westeurope-5.in.applicationinsights.azure.com/;LiveEndpoint=https://westeurope.livediagnostics.monitor.azure.com/;ApplicationId=427c95d5-7252-4513-aed3-e1e5c3eece9d"

File: Packages/OfemKit/Sources/OfemKit/BuildInfo/BuildInfo.swift
import Foundation

/// Runtime identity for the running OFEM binary.
///
/// `version` is the single canonical CalVer source (engine-05).  It is read
/// from `CFBundleShortVersionString` in `Bundle.main` at runtime, so
/// Xcode-built releases automatically pick up the CalVer string
/// (e.g. `"2026.05.1"`) injected by `xcodebuild MARKETING_VERSION=…` on the
/// release tag.  The `"0.0.0-dev"` fallback is returned in unit-test contexts
/// where `Bundle.main` is the `xctest` runner which carries no version key.
///
/// ## Testability (engine-06)
///
/// The bundle read is isolated in ``version(from:)`` so the parsing + fallback
/// logic can be exercised in unit tests against a synthetic `Bundle`.  The
/// top-level ``version`` property is a convenience that calls
/// ``version(from:)`` with `Bundle.main`.
public enum BuildInfo {
    // MARK: - Version

    /// The CalVer release string (e.g. `"2026.05.1"`).
    ///
    /// Reads `CFBundleShortVersionString` from `Bundle.main`. Falls back to
    /// `"0.0.0-dev"` when the key is absent (xctest, command-line tool builds,
    /// or un-versioned debug builds).
    ///
    /// This is the **single canonical version source** for the entire package
    /// (engine-05).  Do not introduce additional version constants.
    public static let version: String = version(from: .main)

    /// Returns the CalVer string from the given bundle, or `"0.0.0-dev"`.
    ///
    /// Exposed for testing so the parse + fallback path can be exercised
    /// with a synthetic `Bundle` (engine-06).
    ///
    /// - Parameter bundle: The bundle to read `CFBundleShortVersionString` from.
    /// - Returns: The version string, or `"0.0.0-dev"` if the key is absent.
    public static func version(from bundle: Bundle) -> String {
        bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString")
            as? String ?? "0.0.0-dev"
    }

    // MARK: - Build timestamp

    /// The ISO-8601 UTC build timestamp injected by the build system
    /// (e.g. `"2026-06-20T14:03:12Z"`).
    ///
    /// Reads `OFEMBuildTimestamp` from `Bundle.main`, which the
    /// `inject-build-timestamp.sh` build phase writes into the compiled
    /// bundle's Info.plist at build time. Returns `nil` when the key is
    /// absent (xctest runner, command-line builds without the script phase).
    public static let buildTimestamp: String? = buildTimestamp(from: .main)

    /// Returns the build timestamp from the given bundle, or `nil`.
    ///
    /// Exposed for testing so the lookup + fallback path can be exercised
    /// with a synthetic `Bundle`.
    ///
    /// - Parameter bundle: The bundle to read `OFEMBuildTimestamp` from.
    /// - Returns: The ISO-8601 timestamp string, or `nil` if the key is absent.
    public static func buildTimestamp(from bundle: Bundle) -> String? {
        bundle.object(forInfoDictionaryKey: "OFEMBuildTimestamp") as? String
    }

    // MARK: - Telemetry

    /// The Application Insights connection string used for opt-out telemetry.
    ///
    /// This is a committed source constant. App Insights connection strings are
    /// write-only by design and ship inside every browser-side JS app, mobile
    /// binary, and desktop app that uses Application Insights. Committing it
    /// here means that source builds and forks report to the same endpoint as
    /// official Homebrew releases, so the maintainer sees a representative signal.
    ///
    /// Users disable telemetry from the menu bar (Settings › Telemetry) or by
    /// setting `OFEM_TELEMETRY=0`. No opt-in path requires editing this file.
    public static let appInsightsConnectionString =
        "InstrumentationKey=bb7c05e2-4616-4b8d-a18a-e32128034eb4;" +
        "IngestionEndpoint=https://westeurope-5.in.applicationinsights.azure.com/;" +
        "LiveEndpoint=https://westeurope.livediagnostics.monitor.azure.com/;" +
        "ApplicationId=427c95d5-7252-4513-aed3-e1e5c3eece9d"
}
